test/jdk/java/net/Socket/CloseOnFailureTest.java

```diff
@@ -0,0 +1,519 @@
+                        connectException.getClass().getSimpleName());
+                return new TestCase(
+                        description,
+                        socketImpl,
+                        caughtException -> assertSame(connectException, caughtException),
+                        () -> assertEquals(expectedCloseInvocationCount, socketImpl.closeInvocationCounter.get()),
+                        _ -> {
+                        });
+            }
+
+        }
+
+        @Override
+        public String toString() {
+            return description;
+        }
+
+    }
+
+    private static abstract class AbstractRealSocketTest {
+
+        private static final InetSocketAddress UNRESOLVED_ADDRESS =
+                InetSocketAddress.createUnresolved("no.such.host", DEAD_SERVER_PORT);
+
+        @Test
+        void verifyUnresolvedAddress() {
+            assertTrue(UNRESOLVED_ADDRESS.isUnresolved());
+        }
+
+        @Test
+        void verifyUnboundSocket() throws IOException {
+            try (Socket socket = createUnboundSocket()) {
+                assertFalse(socket.isBound());
+                assertFalse(socket.isConnected());
+            }
+        }
+
+        @Test
+        void verifyBoundSocket() throws IOException {
+            try (Socket socket = createBoundSocket()) {
+                assertTrue(socket.isBound());
+                assertFalse(socket.isConnected());
+            }
+        }
+
+        @Test
+        @SuppressWarnings("resource")
+        void verifyConnectedSocket() throws IOException {
+            SocketAddress serverSocketAddress = serverSocket().getLocalSocketAddress();
+            try (Socket socket = createConnectedSocket(serverSocketAddress)) {
+                assertTrue(socket.isBound());
+                assertTrue(socket.isConnected());
+            }
+        }
+
+        @Test
+        void socketShouldBeClosedWhenConnectFailsUsingUnboundSocket() throws IOException {
+            try (Socket socket = createUnboundSocket()) {
+                assertThrows(IOException.class, () -> socket.connect(REFUSING_SOCKET_ADDRESS));
+                assertTrue(socket.isClosed());
+            }
+        }
+
+        @Test
+        void socketShouldBeClosedWhenConnectFailsUsingBoundSocket() throws IOException {
+            try (Socket socket = createBoundSocket()) {
+                assertThrows(IOException.class, () -> socket.connect(REFUSING_SOCKET_ADDRESS));
+                assertTrue(socket.isClosed());
+            }
+        }
+
+        @Test
+        @SuppressWarnings("resource")
+        void socketShouldNotBeClosedWhenConnectFailsUsingConnectedSocket() throws IOException {
+            SocketAddress serverSocketAddress = serverSocket().getLocalSocketAddress();
+            try (Socket socket = createConnectedSocket(serverSocketAddress)) {
+                assertReconnectFailure(() -> socket.connect(REFUSING_SOCKET_ADDRESS));
```

Comment:
I needed to delegate the assertion to the implementation because when I call `connect()` on a `Socket` created using `SocketChannel.open(address).socket()` and `new Socket(address, port)`, the former throws an `AlreadyConnectedException`, whereas the latter throws a `SocketException`. Is this discrepancy expected?

Reply:
There are some subtle behaviour discrepancies between plain Socket (created with new Socket()) and the SocketAdaptor (created from SocketChannel.socket()). SocketChannel;;connect will throw AlreadyConnectedException, and so that's what the socket adaptor also does. Ideally we'd want to reduce the distance between plain Socket and the Socket adaptor - but changing behaviours in these APIs is always tricky and a bit risky (and can require a lot of efforts). FWIW - SocketChannel does the right thing - but we're stuck with SocketException in the Socket API due to compatibility reason.

Reply:
Got it. Thanks for the explanation.